external/xtorch_harmonics/xtorch_harmonics/xtorch_harmonics.py

```diff
@@ -0,0 +1,208 @@
+            f"Latitude coordinate {obj[lat_dim]} does not match the expected "
+            f"quadrature points for the provided forward grid {forward_grid!r}."
+        )
+
+    try:
+        if lon_dim in obj.coords:
+            # TODO: I suppose for this check we really only need to ensure that
+            # points are equally spaced and global.  Whether they exactly align
+            # with a regular grid from 0 to 360 degrees E is not necessarily
+            # relevant.
+            np.testing.assert_allclose(obj[lon_dim].data, expected_longitudes)
+            # TODO: could warn or raise if lon_dim not in obj.coords.
+    except AssertionError:
+        raise ValueError(
+            f"Longitude coordinate {obj[lon_dim]} does not match the expected "
+            f"quadrature points for the provided forward grid {forward_grid!r}."
+        )
+
+
+def _roundtrip_numpy(
+    array: np.array, forward_grid: T_Grid, inverse_grid: T_Grid
+) -> np.ndarray:
+    *_, n_lat, n_lon = array.shape
+    tensor = torch.tensor(array).type(torch.double)
+    forward_transform = torch_harmonics.RealSHT(n_lat, n_lon, grid=forward_grid)
+    inverse_transform = torch_harmonics.InverseRealSHT(n_lat, n_lon, grid=inverse_grid)
+    roundtripped = inverse_transform(forward_transform(tensor))
+    return np.array(roundtripped.type(torch.float))
+
+
+def _roundtrip_dataarray(
+    da: xr.DataArray,
+    forward_grid: T_Grid,
+    inverse_grid: T_Grid,
+    lat_dim: Hashable,
+    lon_dim: Hashable,
+) -> xr.DataArray:
+    # Ensure the DataArray is chunked contiguously in the horizontal, which is
+    # required for a 2D spatial transform.
+    horizontal_dims = {lat_dim, lon_dim}
+    da = da.chunk({dim: -1 for dim in horizontal_dims})
+
+    result = xr.apply_ufunc(
+        _roundtrip_numpy,
+        da,
+        input_core_dims=[[lat_dim, lon_dim]],
+        output_core_dims=[[lat_dim, lon_dim]],
+        dask="parallelized",
+        output_dtypes=[np.float32],
+        keep_attrs=True,
+        kwargs={"forward_grid": forward_grid, "inverse_grid": inverse_grid},
+    )
+
+    # Restore dimension order to match input DataArray
+    return result.transpose(*da.dims)
+
+
+def _roundtrip_dataset(
+    ds: xr.Dataset,
+    forward_grid: T_Grid,
+    inverse_grid: T_Grid,
+    lat_dim: Hashable,
+    lon_dim: Hashable,
+) -> xr.Dataset:
+    horizontal_dims = {lon_dim, lat_dim}
+    results = []
+    for da in ds.data_vars.values():
+        if not horizontal_dims.issubset(da.dims):
+            results.append(da)
+        else:
+            result = _roundtrip_dataarray(
+                da, forward_grid, inverse_grid, lat_dim, lon_dim
+            )
+            results.append(result)
+    return xr.merge(results).assign_attrs(ds.attrs)
+
+
+def roundtrip(
+    obj: T_XarrayObject,
+    forward_grid: T_Grid = LEGENDRE_GAUSS_GRID,
+    inverse_grid: T_Grid = LEGENDRE_GAUSS_GRID,
+    lat_dim: Hashable = LAT_DIM,
+    lon_dim: Hashable = LON_DIM,
```

Comment:
nit: minor preference to not have defaults for these args, since this package is very general purpose.

Reply:
Yeah, that's probably better.  The coarsening functions can be awkward to deal with sometimes because of this.